File: 2016/16 Dragon Checksum/solution.py
input = seed = "11101000110010100"
N = 272
#N = 35651584

#input = "10000"
#N = 20
while len(input) < N:
    input = (input + "0" + input[::-1].replace("0", ".").replace("1", "0").replace(".", "1"))[:N]

while True:
    input = "".join(str(int(x==y)) for x, y in zip(input[::2], input[1::2]))
    print(input)
    if len(input) % 2:
        break

s = []
for i in range(N):
    rep = i // (len(seed) + 1)
    inre = i % (len(seed) + 1)
    if inre == len(seed):
        n = rep + 1
        c = str(int((((n & -n) << 1) & n) != 0))
    elif rep % 2 == 0:
        c = seed[inre]
    else:
        c = str(1 - int(seed[-inre-1]))
    s.append(c)
    print(c, end="")
print()

lb = N & -N
res = ""
t = iter(s)
for i in range(N // lb):
    st = []
    for j in range(1, 1 + lb):
        st.append(next(t))
        print("v", end="")
        while not j & 1:
            print("z", end="")
            st.append(int(st.pop() == st.pop()))
            j >>= 1
    res += str(st[0])
print()
print(res)
#print("".join(str(int((((n & -n) << 1) & n) != 0)) for n in range(1, N + 1)))

File: 2016/16 Dragon Checksum/solution.cpp
#include <stdio.h>
#include <string.h>

char seed[] = "11101000110010100";
char seedlen = strlen(seed);

char dragon(long i) {
    long rep = i / (seedlen + 1);
    long inre = i % (seedlen + 1);
    if (inre == seedlen) {
        rep++;
        return (((rep & -rep) << 1) & rep) != 0;
    }
    else if (rep & 1)
        return '1' - seed[seedlen - inre - 1];
    else
        return seed[inre] - '0';
}

void compute(long n, char *res) {
    char stack[32], *ptr;
    long chunk_size = n & -n;
    printf("cd %li\n", chunk_size);
    for(long chunks = n / chunk_size, counter = 0; chunks--;) {
        ptr = stack;
        for(long i = 1; i <= chunk_size; i++) {
            *ptr++ = dragon(counter++);
            for(long j = i; !(j & 1); j >>= 1) {
                char p1 = *--ptr;
                char p2 = *--ptr;
                *ptr++ = (p1 == p2);
            }
        }
        *res++ = '0' + *stack;
    }
    *res = 0;
    printf("\n");
}

char res1[32], res2[32];

int main() {
    compute(272, res1);
    printf("%s\n", res1);

    compute(35651584, res2);
    printf("%s\n", res2);
}
